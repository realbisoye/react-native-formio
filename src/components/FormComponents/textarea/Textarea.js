
import React from 'react';
import {TextInput} from 'react-native';
import InputComponent from '../sharedComponents/Input';
import styles from './styles';

export default class Textarea extends InputComponent {

  constructor(props) {
    super(props);
    this.onChangeText = this.onChangeText.bind(this);
    this.getSingleElement = this.getSingleElement.bind(this);
  }

  onChangeText(index, value) {
    this.setValue(value, index);
  }

  getSingleElement(value, index) {
    const {component, name, readOnly, colors, theme} = this.props;
    const fieldValue = typeof value === 'object' ? value.item : value;
    index = index || 0;
    return (
      <TextInput
        key={index}
        id={component.key}
        data-index={index}
        name={name}
        value={fieldValue}
        defaultValue={fieldValue}
        style={[styles.textarea, {borderColor: colors.borderColor, lineHeight: theme.Input.lineHeight}]}
        multiline
        autoCorrect
        numberOfLines={component.rows}
        disabled={!readOnly}
        placeholder={component.placeholder}
        onChange={this.onChange}
      />
    );
  }
}
