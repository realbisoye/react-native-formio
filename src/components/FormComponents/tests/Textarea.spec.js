import React from 'react';
import renderer from 'react-test-renderer';
import colors from '../../../defaultTheme/colors';
import theme from '../../../defaultTheme';
import Textarea from '../textarea/Textarea';

 describe('Textarea', () => {
  describe('Single Textarea', () => {
    const component= {
      'input': true,
      'tableView': true,
      'label': 'textarea',
      'key': 'textarea',
      'placeholder': '',
      'prefix': '',
      'suffix': '',
      'rows': 3,
      'multiple': false,
      'defaultValue': '',
      'protected': false,
      'persistent': true,
      'wysiwyg': false,
      'validate': {
        'required': false,
        'minLength': '',
        'maxLength': '',
        'pattern': '',
        'custom': ''
      },
      'type': 'textarea',
      'conditional': {
        'show': '',
        'when': null,
        'eq': ''
      }
    };
    const attachToForm = jest.fn();
     it('Renders a basic textarea', () => {
      const element = renderer.create(
        <Textarea
          component={component}
          attachToForm={attachToForm}
          colors={colors}
          theme={theme}
        />
      );
      expect(element).toMatchSnapshot();
    });

     it.skip('Fills in the placeholder value', () => {
      component.placeholder = 'My Placeholder';
      const element = renderer.create(
        <Textarea
          component={component}
          attachToForm={attachToForm}
          colors={colors}
          theme={theme}
        />
      ).find('textarea');
      expect(element.attr('placeholder')).to.equal('My Placeholder');
      component.placeholder = '';
    });
     it.skip('Renders with a prefix', () => {
      component.prefix = '$';
      const element = renderer.create(
        <Textarea
          component={component}
          attachToForm={attachToForm}
          colors={colors}
          theme={theme}
        />
      ).find('.input-group');
      expect(element.children().length).to.equal(2);
      expect(element.children().eq(0).hasClass('input-group-addon')).to.equal(true);
      expect(element.children().eq(0).html()).to.equal('$');
      component.prefix = '';
    });
     it.skip('Renders with a suffix', () => {
      component.suffix = 'Pounds';
      const element = renderer.create(
        <Textarea
          component={component}
          attachToForm={attachToForm}
          colors={colors}
          theme={theme}
        />
      ).find('.input-group');
      expect(element.children().length).to.equal(2);
      expect(element.children().eq(1).hasClass('input-group-addon')).to.equal(true);
      expect(element.children().eq(1).html()).to.equal('Pounds');
      component.suffix = '';
    });
     it.skip('Renders with prefix and suffix', () => {
      component.prefix = 'Prefix';
      component.suffix = 'Suffix';
      const element = renderer.create(
        <Textarea
          component={component}
          attachToForm={attachToForm}
          colors={colors}
          theme={theme}
        />
      ).find('.input-group');
      expect(element.children().length).to.equal(3);
      expect(element.children().eq(0).hasClass('input-group-addon')).to.equal(true);
      expect(element.children().eq(0).html()).to.equal('Prefix');
      expect(element.children().eq(2).hasClass('input-group-addon')).to.equal(true);
      expect(element.children().eq(2).html()).to.equal('Suffix');
      component.prefix = '';
      component.suffix = '';
    });
     it.skip('Check single textarea with required', () => {
      component.validate.required = true;
      const element = renderer.create(
        <Textarea
          component={component}
          attachToForm={attachToForm}
          colors={colors}
          theme={theme}
        />
      );
      expect(element.find('.formio-component-single label ').attr('class')).to.equal('control-label field-required');
    });
     it.skip('Check single textarea without required', () => {
      component.validate.required = false;
      const element = renderer.create(
        <Textarea
          component={component}
          attachToForm={attachToForm}
          colors={colors}
          theme={theme}
        />
      );
      expect(element.find('.formio-component-single label ').attr('class')).to.equal('control-label');
    });
     it.skip('Check single textarea without label', () => {
      component.label = '';
      const element = renderer.create(
        <Textarea
          component={component}
          attachToForm={attachToForm}
          colors={colors}
          theme={theme}
        />
      );
      expect(element.find('.formio-component-single label').length).to.equal(0);
    });
     it.skip('sets a custom class', () => {
      component.customClass = 'my-custom-class';
      const element = renderer.create(
        <Textarea
          component={component}
          attachToForm={attachToForm}
          colors={colors}
          theme={theme}
        />
      ).children().eq(0);
      expect(element.attr('class').split(' ')).to.contain('my-custom-class');
    });
   });
   describe('Multiple Textarea', () => {
    const component = {
      'input': true,
      'tableView': true,
      'label': 'textarea',
      'key': 'textarea',
      'placeholder': '',
      'prefix': '',
      'suffix': '',
      'rows': 3,
      'multiple': true,
      'defaultValue': '',
      'protected': false,
      'persistent': true,
      'wysiwyg': false,
      'validate': {
        'required': false,
        'minLength': '',
        'maxLength': '',
        'pattern': '',
        'custom': ''
      },
      'type': 'textarea',
      'conditional': {
        'show': '',
        'when': null,
        'eq': ''
      }
    };
    const attachToForm = jest.fn();
     it.skip('Renders a multi-value textarea', () => {
      const element = renderer.create(
        <Textarea
          name='textarea'
          component={component}
          attachToForm={attachToForm}
          colors={colors}
          theme={theme}
        />
      ).find('.form-field-type-textarea');
      expect(element).to.have.length(1);
      expect(element.hasClass('form-group form-field-type-textarea form-group-textarea')).to.equal(true);
      expect(element.attr('id')).to.equal('form-group-textarea');
      expect(element.children().eq(0).hasClass('formio-component-multiple')).to.equal(true);
      expect(element.children().eq(0).children().eq(0).attr('for')).to.equal('textarea');
      expect(element.children().eq(0).children().eq(0).hasClass('control-label')).to.equal(true);
      expect(element.children().eq(0).children().eq(0).text()).to.equal('textarea');
      const table = element.children().eq(0).children().eq(1);
      expect(table.hasClass('table table-bordered')).to.equal(true);
      expect(table.find('tr').length).to.equal(2);
      expect(table.find('tr td div.input-group').length).to.equal(1);
      expect(table.find('tr td div.input-group textarea').attr('placeholder')).to.equal('');
      //expect(table.find('tr td div.input-group textarea').attr('value')).to.equal('');
      expect(table.find('tr td div.input-group textarea').attr('id')).to.equal('textarea');
      expect(table.find('tr td div.input-group textarea').attr('name')).to.equal('textarea');
      expect(table.find('tr td div.input-group textarea').attr('class')).to.equal('form-control');
      expect(table.find('tr td div.input-group textarea').attr('data-index')).to.equal('0');
    });
     it.skip('Fills in the placeholder value', () => {
      component.placeholder = 'My Placeholder';
      const element = renderer.create(
        <Textarea
          name='textarea'
          component={component}
          attachToForm={attachToForm}
          colors={colors}
          theme={theme}
        />
      ).find('textarea#textarea');
      expect(element.attr('placeholder')).to.equal('My Placeholder');
      component.placeholder = '';
    });
     it.skip('Renders with a prefix', () => {
      component.prefix = '$';
      const element = renderer.create(
        <Textarea
          name='textarea'
          component={component}
          attachToForm={attachToForm}
          colors={colors}
          theme={theme}
        />
      ).find('.input-group');
      expect(element.children().length).to.equal(2);
      expect(element.children().eq(0).hasClass('input-group-addon')).to.equal(true);
      expect(element.children().eq(0).html()).to.equal('$');
      component.prefix = '';
    });
     it.skip('Renders with a suffix', () => {
      component.suffix = 'Pounds';
      const element = renderer.create(
        <Textarea
          name='textarea'
          component={component}
          attachToForm={attachToForm}
          colors={colors}
          theme={theme}
        />
      ).find('.input-group');
      expect(element.children().length).to.equal(2);
      expect(element.children().eq(1).hasClass('input-group-addon')).to.equal(true);
      expect(element.children().eq(1).html()).to.equal('Pounds');
      component.suffix = '';
    });
     it.skip('Renders with prefix and suffix', () => {
      component.prefix = 'Prefix';
      component.suffix = 'Suffix';
      const element = renderer.create(
        <Textarea
          name='textarea'
          component={component}
          attachToForm={attachToForm}
          colors={colors}
          theme={theme}
        />
      ).find('.input-group');
      expect(element.children().length).to.equal(3);
      expect(element.children().eq(0).hasClass('input-group-addon')).to.equal(true);
      expect(element.children().eq(0).html()).to.equal('Prefix');
      expect(element.children().eq(2).hasClass('input-group-addon')).to.equal(true);
      expect(element.children().eq(2).html()).to.equal('Suffix');
      component.prefix = '';
      component.suffix = '';
    });
     it.skip('Adds and removes rows', () => {
      component.defaultValue = 'My Value';
      const element = renderer.create(
        <Textarea
          name='textarea'
          component={component}
          attachToForm={attachToForm}
          colors={colors}
          theme={theme}
        />
      ).find('.form-field-type-textarea');
      const table = element.find('table');
      table.find('a.btn.add-row').simulate('click');
      expect(table.find('tr').length).to.equal(3);
      expect(table.find('tr').at(0).find('textarea').prop('data-index')).to.equal(0);
      expect(table.find('tr').at(1).find('textarea').prop('data-index')).to.equal(1);
      table.find('a.btn.add-row').simulate('click');
      expect(table.find('tr').length).to.equal(4);
      expect(table.find('tr').at(0).find('textarea').prop('data-index')).to.equal(0);
      expect(table.find('tr').at(1).find('textarea').prop('data-index')).to.equal(1);
      expect(table.find('tr').at(2).find('textarea').prop('data-index')).to.equal(2);
      table.find('a.btn.add-row').simulate('click');
      expect(table.find('tr').length).to.equal(5);
      expect(table.find('tr').at(0).find('textarea').prop('data-index')).to.equal(0);
      expect(table.find('tr').at(1).find('textarea').prop('data-index')).to.equal(1);
      expect(table.find('tr').at(2).find('textarea').prop('data-index')).to.equal(2);
      expect(table.find('tr').at(3).find('textarea').prop('data-index')).to.equal(3);
      table.find('a.btn.remove-row-3').simulate('click');
      expect(table.find('tr').length).to.equal(4);
      expect(table.find('tr').at(0).find('textarea').prop('data-index')).to.equal(0);
      expect(table.find('tr').at(1).find('textarea').prop('data-index')).to.equal(1);
      expect(table.find('tr').at(2).find('textarea').prop('data-index')).to.equal(2);
      table.find('a.btn.remove-row-1').simulate('click');
      expect(table.find('tr').length).to.equal(3);
      expect(table.find('tr').at(0).find('textarea').prop('data-index')).to.equal(0);
      expect(table.find('tr').at(1).find('textarea').prop('data-index')).to.equal(1);
      table.find('a.btn.remove-row-1').simulate('click');
      expect(table.find('tr').length).to.equal(2);
      expect(table.find('tr').at(0).find('textarea').prop('data-index')).to.equal(0);
      table.find('a.btn.remove-row-0').simulate('click');
      expect(table.find('tr').length).to.equal(1);
      table.find('a.btn.add-row').simulate('click');
      expect(table.find('tr').length).to.equal(2);
      expect(table.find('tr').at(0).find('textarea').prop('data-index')).to.equal(0);
    });
     it.skip('Check multiple textarea with required', () => {
      component.validate.required = true;
      const element = renderer.create(
        <Textarea
          component={component}
          attachToForm={attachToForm}
          colors={colors}
          theme={theme}
        />
      );
      expect(element.find('.formio-component-multiple label ').attr('class')).to.equal('control-label field-required');
    });
     it.skip('Check multiple textarea without required', () => {
      component.validate.required = false;
      const element = renderer.create(
        <Textarea
          component={component}
          attachToForm={attachToForm}
          colors={colors}
          theme={theme}
        />
      );
      expect(element.find('.formio-component-multiple label ').attr('class')).to.equal('control-label');
    });
     it.skip('Check multiple textarea without label', () => {
      component.label = '';
      const element = renderer.create(
        <Textarea
          component={component}
          attachToForm={attachToForm}
          colors={colors}
          theme={theme}
        />
      );
      expect(element.find('.formio-component-multiple label').length).to.equal(0);
    });
     it.skip('sets a custom class', () => {
      component.customClass = 'my-custom-class';
      const element = renderer.create(
        <Textarea
          component={component}
          attachToForm={attachToForm}
          colors={colors}
          theme={theme}
        />
      ).children().eq(0);
      expect(element.attr('class').split(' ')).to.contain('my-custom-class');
    });
  });
});
