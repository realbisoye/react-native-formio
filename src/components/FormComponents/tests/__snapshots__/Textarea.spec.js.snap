// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Textarea Single Textarea Renders a basic textarea 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Array [
        Object {},
        undefined,
      ]
    }
  >
    <Text
      style={
        Array [
          Object {},
          undefined,
          undefined,
          undefined,
          undefined,
          undefined,
          undefined,
          undefined,
          undefined,
          undefined,
          Array [
            Object {
              "color": "#86939e",
              "fontSize": 15,
              "fontWeight": "bold",
              "marginBottom": 1,
              "marginLeft": 20,
              "marginRight": 20,
              "marginTop": 15,
            },
            undefined,
            undefined,
          ],
        ]
      }
    >
      <Text
        style={
          Array [
            Object {},
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
          ]
        }
      >
        
      </Text>
       
      <Text
        style={
          Array [
            Object {},
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
          ]
        }
      >
        
      </Text>
       
      textarea
       
      <Text
        style={
          Array [
            Object {},
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
          ]
        }
      >
        
      </Text>
    </Text>
  </View>
  <TextInput
    allowFontScaling={true}
    autoCorrect={true}
    data-index={0}
    defaultValue=""
    disabled={true}
    id="textarea"
    multiline={true}
    numberOfLines={3}
    onChange={[Function]}
    placeholder=""
    style={
      Array [
        Object {
          "borderWidth": 1,
          "fontSize": 16,
          "lineHeight": 16,
          "marginHorizontal": 10,
          "marginTop": 20,
          "paddingBottom": 20,
          "paddingHorizontal": 10,
          "paddingTop": 10,
        },
        Object {
          "borderColor": "#9e9e9e",
          "lineHeight": 18,
        },
      ]
    }
    underlineColorAndroid="transparent"
    value=""
  />
</View>
`;
